Kinematics only

woensug-choi opened the pull request:
New kinematics only PR

woensug-choi committed:
initiation of the new branch

woensug-choi committed:
basic set-up first init

woensug-choi committed:
initial templates ready for commanding with frl_vehicle_msgs

woensug-choi committed:
all working independently, however, pitch and rudder control stop the vehicle. (subscribing function needs check)

woensug-choi committed:
Commanding function ready

woensug-choi committed:
UwGliderStatus puglisher for depth and RPY added

woensug-choi committed:
Merge branch 'DirectROSKinematics_plugin' of github.com:Field-Robotics-Lab/glider_hybrid_whoi into DirectROSKinematics_plugin

woensug-choi committed:
Added DVL sensor and new worlds for altitude status

woensug-choi committed:
connected gps sensor data

woensug-choi committed:
fixed minor sign errors

woensug-choi committed:
initial buoyancy engine added (cCOB=COG)

woensug-choi committed:
pumped volume unit fix

woensug-choi committed:
added commands with target_pumped_volume

woensug-choi committed:
visual effects for propeller rotation added

woensug-choi committed:
dynamics added

woensug-choi committed:
dynamics on if pitch control type is battery position

woensug-choi committed:
Iter 1, Debug trial 1

woensug-choi committed:
buoyancy fixed, thruster changed to using AddRelativeForee, Rudder control excluded

woensug-choi committed:
buoyancy sign fixed for dynamics

woensug-choi committed:
Temporarily restricted forces in dynamics to 3 dimension only for stability

woensug-choi committed:
pitch_cmd_type: 0 option reconsidered

woensug-choi committed:
added a function for our thruster (voltage input needs investigation)

woensug-choi committed:
multiply 2.0 for thruster power value (dual thrusters)

woensug-choi committed:
free surface detection added

daewok committed:
Improve interface diagram

Add slocum_glider_sim_driver and slocum_glider_sim_console nodes. Make clear
that the "backseat driver" interacts with the sim driver via its extctl
interface. Provide link to where these nodes are located.

woensug-choi committed:
Updating plugin diagram

woensug-choi committed:
Usable kinematics

woensug-choi committed:
Merge branch 'DirectROSKinematics_plugin' of github.com:Field-Robotics-Lab/glider_hybrid_whoi into DirectROSKinematics_plugin

woensug-choi committed:
Usable Kinematics functionality

woensug-choi committed:
surface detection for kinematics fixed

woensug-choi committed:
Surface detection fix, heading angle fix, switch statement conversion

daewok committed:
In the kinematic case, servoing on pitch is equivalent to setting it once

woensug-choi committed:
add ci check

woensug-choi committed:
Merge branch 'DirectROSKinematics_plugin' of https://github.com/Field-Robotics-Lab/glider_hybrid_whoi into DirectROSKinematics_plugin

woensug-choi committed:
fix ci check

woensug-choi committed:
add code_check script

woensug-choi committed:
remove ci check, uuv_simulator inhereted code cause tons of problems

woensug-choi committed:
add sat_nav_fix.msg

woensug-choi committed:
Merge branch 'DirectROSKinematics_plugin' of github.com:Field-Robotics-Lab/glider_hybrid_whoi into DirectROSKinematics_plugin

woensug-choi committed:
initial commit for kinematics only branch

woensug-choi committed:
remove urdf test

woensug-choi committed:
docker folder added

woensug-choi committed:
Introducing docker

woensug-choi committed:
thruster coefficient definable at sdf

woensug-choi committed:
Update README.md

New README for kinematics

woensug-choi committed:
double thruster power already defined at sub function

woensug-choi committed:
Update README.md

woensug-choi commented:
# Kinematics plugin for WHOI hybrid gliders
Kinematics control plugin for WHOI hybrid gliders

## How-to
### Installation
* First check to make sure you meet the [System Requirements](https://github.com/Field-Robotics-Lab/dave/wiki/System-Requirements).
* Then choose from one of the following two installation options:
    1. **Directly on Host**
         1. [Install environment and dependent repositories](https://github.com/Field-Robotics-Lab/dave/wiki/Install-Directly-on-Host) : Instructions to install ROS Melodic, Gazebo 9, UUV Simulator and DAVE directly on your host machine.
         2. Clone this repository in `~/uuv/src` folder and compile with `catkin_make` at `~/uuv` directory.
    2. **Using Docker**
         1. Make sure you have Docker v19.03 or higher ([installation instructions](https://docs.docker.com/engine/install/ubuntu/)) and nvidia-container-toolkit ([installation instructions](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html#setting-up-nvidia-container-toolkit))
         2. Run the `build.bash` script located in the `docker` folder of this repository
             ```
             ./build.bash .
             ```
         3. Run the container with `run.bash` script
             ```
             ./run.bash glider_kinematics
             ```
         * Opening additional terminals
             ```
             ./join.bash glider_kinematics
             ```
### Quickstart
* Running the simulator (Run each commands in separate terminal window)
    1. Spawn underwater world with gazebo
        ```
        roslaunch glider_hybrid_whoi_gazebo seafloor_underwater_world.launch
        ```
    2. Spawn glider into the gazebo world
        ```
        roslaunch glider_hybrid_whoi_gazebo start_demo_teleop_kinematics.launch joy_id:=0
        ``` 
    3. Control glider with ROS
        ```
        # Using test python script
        rosrun direct_kinematics_ros_plugins test_directKinematics.py
        # Directly to the ros node
        # Toward surface
        rostopic pub /glider_hybrid_whoi/direct_kinematics/UwGlidCommand frl_vehicle_msgs/UwGliderCommand '{motor_cmd_type: 1, target_motor_cmd: 2, pitch_cmd_type: 2, target_pitch_value: -0.4, rudder_control_mode: 1, target_heading: 0.8}'
        # Back to underwater
        rostopic pub /glider_hybrid_whoi/direct_kinematics/UwGlidCommand frl_vehicle_msgs/UwGliderCommand '{motor_cmd_type: 1, target_motor_cmd: 2, pitch_cmd_type: 2, target_pitch_value: 0.4, rudder_control_mode: 1, target_heading: -0.8}'
        ```
    4. Glider status
        ```
        rostopic echo /glider_hybrid_whoi/direct_kinematics/UwGliderStatus
        ```

## Features
* Organized based on support for [frl_vehicle_msgs](https://github.com/Field-Robotics-Lab/frl_msgs/tree/master/frl_vehicle_msgs/msg)
### Glider dynamic parameters
- Parametes for pitch control, buoyancy induced velocity with the flight model, and thruster power is defined at [glider_hybrid_whoi_base_kinematics.xacro](https://github.com/Field-Robotics-Lab/glider_hybrid_whoi/blob/10524388cce32865ae051e285dbe631ea89159e4/glider_hybrid_whoi_description/urdf/glider_hybrid_whoi_base_kinematics.xacro#L139)
#### Pitch control
| Battery position | Target once | Target servo |
| ------------- | ------------- | ------------- |
| :heavy_check_mark:  | :heavy_check_mark:  | :heavy_check_mark: |
- Change of pitch is defined with first-order function with two coefficients (mx + b) for battery position command values
- coefficient name : `f_pitch_battpos_cal_m`, `f_pitch_battpos_cal_b`

#### Thruster control
| Voltage | Power |
| ------------- | ------------- | 
| :heavy_check_mark:  | :heavy_check_mark: |
- Thruster power is defined with 2nd-order fucntion with three coefficients (ax^2 + bx + c) for both voltage/power command values
- coefficient name : `f_thruster_voltage_v1`, `f_thruster_voltage_v2`, `f_thruster_voltage_v3`, `f_thruster_power_w1`, `f_thruster_power_w2`, `f_thruster_power_w3`
- Final thrust to the glider is calculated with `sqrt(motorpower/(fluidDensity*Area*C_D)`
- Propeller rotation is visualized using the command value as rotation speed.

#### Rudder control
| Heading | Angle (Center, Port, Stbd) | Angle (Direct) |
| ------------- | ------------- | ------------- | 
| :heavy_check_mark:  | :heavy_check_mark: | :heavy_check_mark: |
- The size of the port and starboard angle is set to PI/6.

#### Buoyancy engine control
| Flight model |
| ------------- |
| :heavy_check_mark:  |
- Simple flight model is adopted to change the angle of attach with pumped volume to produce buoyancy induced velocity
- +/- 3 degrees depending on the sign of the pitch
- Calculated with Equations 24 and 28 from Eichhorn et al. paper.

#### Glider status
| Lat/Lon | Roll-Pitch-Yaw | Heading | Depth | Altitude | Power | Rudder angle | Battery position | Pumped Volume | nav_sat_fix | 
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| :heavy_check_mark:*  | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark:* | :heavy_multiplication_x: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
- Lat/Lon is acquired from the GPS sensor and Altitude from the DVL sensor. Which is also sent through nav_sat_fix

#### Code Structure Diagram
![image](https://user-images.githubusercontent.com/7955120/101485884-2ba8d400-399f-11eb-90ab-6f1be48d3f18.png)

woensug-choi commented:
#### To-dos for First half of December
- [x] Strip-off dynamics part and make new Kinematics PR (including Docker)
- [x] Add nav_sat_fix to glider status (https://github.com/Field-Robotics-Lab/frl_msgs/pull/8)
- [x] Standalone case-set PR with Buzzard Bay bathymetry & Stratified ocean current (with Kolumbo data)

woensug-choi commented:
@daewok Dockerfile is at `docker` folder :+1: It takes quite some time to build. Check [installation with Docker](https://github.com/Field-Robotics-Lab/glider_hybrid_whoi/tree/Kinematics_only#installation) section how to use `build.bash`, `run.bash`, `join.bash`

daewok commented on README.md:
```diff
@@ -1,31 +1,91 @@
+#### Pitch control
+| Battery position | Target once | Target servo |
+| ------------- | ------------- | ------------- |
+| :heavy_check_mark:  | :heavy_check_mark:  | :heavy_multiplication_x: |
```
Target servo works as well: https://github.com/Field-Robotics-Lab/glider_hybrid_whoi/blob/Kinematics_only/direct_kinematics_ros_plugins/src/DirectKinematicsROSPlugin.cc#L385

daewok commented on docker/Dockerfile:
```diff
@@ -0,0 +1,136 @@
+# Ubuntu 18.04 with nvidia-docker2 beta opengl support.
+ARG BASEIMG=nvidia/opengl:1.0-glvnd-devel-ubuntu18.04
+FROM $BASEIMG
```
Is using this as the base image *really* necessary? If we can use one of the ROS images, it would cut down on build time and disk usage quite a bit. According to the [ROS wiki](https://wiki.ros.org/docker/Tutorials/Hardware%20Acceleration), libglvnd should already be included in their images starting with Melodic.

The only downsides to using the ROS images that I'm aware of is that the ones in the "Official Images" program (i.e., `ros` with no namespace prefix) are up to a month behind on ROS updates (they can't get their act together to make infrastructure that automatically updates the images when they make releases) and they don't publish the desktop-full variant there, so we'd have to use something like `ros:melodic-ros-base` instead. (That won't be an issue, however, if you accept my rosdep suggestion below)

They do publish the desktop-full variant under `osrf/ros:melodic-desktop-full`, but those images are (currently) four months old and seem to only be updated when enough people complain about them, so I wouldn't recommend using them.

woensug-choi replied:
I inherited the basic structure of Dockerfile from the Dave project. I thought It's using the base image for the gazebo GUI to work. Changing it wouldn't be a problem for me as long as it works. Do you think changing the base image to `ros:melodic-ros-base` wouldn't cause any problems for the gazebo? Or may be Nvidia image without libglvnd included at Dockerhub (https://hub.docker.com/r/nvidia/opengl)?

daewok commented on docker/Dockerfile:
```diff
@@ -0,0 +1,136 @@
+        libeigen3-dev \
+        pkg-config \
+        protobuf-compiler \
+ && apt clean
```
It's also standard to `rm -rf /var/lib/apt/lists/*` in any RUN step that does `apt update`

daewok commented on docker/Dockerfile:
```diff
@@ -0,0 +1,136 @@
+ARG DIST=melodic
+
+# Set Gazebo verison
+ARG GAZ=gazebo9
```
This is unused and I don't think it should be used. ROS will automatically pull in the correct version of Gazebo

daewok commented on docker/Dockerfile:
```diff
@@ -0,0 +1,136 @@
+    ros-${DIST}-robot-localization \
+    ros-${DIST}-robot-state-publisher \
+    ros-${DIST}-xacro \
+    libignition-math2 \
```
This installs way more than is needed and requires that if a new dependency is added to a package that we modify the Dockerfile as well. I'd recommend using a multi-stage build where the dependencies are determined using `rosdep` in the first stage and the next stage installs them. See https://gitlab.com/mit-mers/ros/slocum_glider/-/merge_requests/9 for an example (I was just inspired to make this change to our Docker image).

woensug-choi replied:
Adopted install_script methods from your gitlab dockerfile

daewok commented on docker/Dockerfile:
```diff
@@ -0,0 +1,136 @@
+# Add a user with the same user_id as the user outside the container
+# Requires a docker build argument `user_id`.
+
+RUN curl -SsL https://github.com/boxboat/fixuid/releases/download/v0.4/fixuid-0.4-linux-amd64.tar.gz | tar -C /usr/local/bin -xzf - && \
```
This is cool. I've done similar things using my own hacky scripts without realizing fixuid exists... Gonna have to start using it myself.

daewok commented on docker/Dockerfile:
```diff
@@ -0,0 +1,136 @@
+RUN mkdir -p uuv_ws/src/
+
+# Source all the needed environment files.
+RUN /bin/sh -c 'echo ". /opt/ros/${DIST}/setup.bash" >> ~/.bashrc' \
+ && /bin/sh -c 'echo ". /usr/share/gazebo/setup.sh" >> ~/.bashrc'
+
+ENTRYPOINT ["fixuid"]
```
I'd get rid of the modifications to .bashrc (and the one near the bottom of the file). Instead create a shell script that sources `/home/developer/uuv_ws/devel/setup.bash` and ends with `exec fixuid "$@"`. Then use this script as the entrypoint.

As it currently stands, this image is really only usable for launching an interactive bash shell (i.e., I don't think we can use this with docker-compose without gross kludges, such as setting the command to `bash -c ". /home/developer/uuv_ws/devel/setup.bash && roslaunch ..."`). With the suggested changes, any command can be used to start the container and it will have access to all the ROS envvars.

daewok replied:
Actually, it may be worth it to leave the .bashrc modifications as well. That way if join.bash is used to start a shell, it'll also have access to the ROS environment.

woensug-choi replied:
Can you be more specific about this? ending with `exec fixuid "$@"`?

daewok replied:
See the commits I just pushed to the new branch

daewok commented on docker/Dockerfile:
```diff
@@ -0,0 +1,136 @@
+RUN git clone https://github.com/Field-Robotics-Lab/dave.git
+RUN git clone https://github.com/Field-Robotics-Lab/nps_uw_sensors_gazebo.git
+RUN git clone https://github.com/Field-Robotics-Lab/frl_msgs.git
+RUN git clone https://github.com/uuvsimulator/uuv_simulator.git
```
This is released and available through apt. We should install it through apt (preferrably via `rosdep`) instead of cloning it unless we really need the development version.

woensug-choi replied:
Referenced the method from gitlab dockerfile and changed to download release tarballs if available

daewok commented on docker/Dockerfile:
```diff
@@ -0,0 +1,136 @@
+
+# Git clone required repositories
+WORKDIR /home/developer/uuv_ws/src
+RUN git clone https://github.com/Field-Robotics-Lab/dave.git
+RUN git clone https://github.com/Field-Robotics-Lab/nps_uw_sensors_gazebo.git
+RUN git clone https://github.com/Field-Robotics-Lab/frl_msgs.git
+RUN git clone https://github.com/uuvsimulator/uuv_simulator.git
+RUN git clone https://bitbucket.org/whoidsl/ds_sim.git
+RUN git clone https://bitbucket.org/whoidsl/ds_msgs.git
+RUN git clone https://github.com/Field-Robotics-Lab/glider_hybrid_whoi.git
```
This is 100% not what we want to do. Docker is very aggressive about caching build stages. That means if nothing above these lines change (I don't see how they would, except for a new version of the base image being released), Docker won't even run these commands and we won't get any new commits.

It looks like all the FRL packages (except frl_msgs) have release tarballs. I'd recommend recording version numbers for them (as either ENV steps or in a text file that is ADDed) and downloading the tarballs instead of using git. If there's something we need a development version for like some unreleased feature of DAVE (or any of the WHOI packages... I don't think they have release tarballs) then we should record a git commit in the same way. That way changing any of those versions or commits acts as a cache buster.

woensug-choi replied:
Referenced the method from gitlab dockerfile and changed to download release tarballs if available

daewok commented on docker/Dockerfile:
```diff
@@ -0,0 +1,136 @@
+
+# Update and compile git repositories
+WORKDIR /home/developer/uuv_ws/src
+RUN /bin/sh -c 'ls | xargs -I{} git -C {} pull'
```
Again, this does nothing because of Docker's layer caching.

daewok commented on docker/Dockerfile:
```diff
@@ -0,0 +1,136 @@
+RUN /bin/bash -c 'pip install pandas'
+
+# Set environment
+RUN /bin/sh -c 'echo "source ~/uuv_ws/devel/setup.bash" >> ~/.bashrc'
```
See above comment about removing this.

daewok reviewed (commented):
I mostly did a Dockerizing review. If you want a hand in changing the Dockerfile to respond to anything I wrote, just let me know and I'm happy to make some commits or have a meeting with you after the 20th (life is too crazy to do much on this before then, but asynchronous text like this is mostly fine).

I like to think that there are generally two types of Docker images: development and deployment images. The Dave Dockerfile you based this off of is definitely a development image that just adds all the dependencies but no source code, to make for easy editing and development. The MERS Dockerfile is more of a deployment image (and I should really make a development image for it...). What you've got here seems to try and split the difference which IMO makes it harder to use overall.

I'd suggest either picking one or doing both as separate images. If you decide to go more the development image route, I'd still install the dependencies inside the image, just in a separate workspace that you then overlay the workspace for this repo on top of. The reason for that is if I want to develop this project in Docker I don't want to have to check out the correct versions of all the deps if I only want to change a file in this project.

In the long run, I'd also start pressuring the maintainers of those WHOI bitbucket repos to do actual releases as well.

woensug-choi commented:
@daewok I've referenced as much as I can from your GitLab Dockefile. I've made it through catkin_make compile. But was not able to play the plugin.
1. Gazebo doesn't launch with complaints of GUI driver fail. I tried to add `nvidia/opengl:base-ubuntu18.04` at runtime image but without success. (Maybe `run.bash` issue?)
https://github.com/Field-Robotics-Lab/glider_hybrid_whoi/blob/c1b3aff1b45be5935b931e32933af7dec21f5b8d/docker/Dockerfile#L95-L96
2. I thought the method with install-script would install dependent packages (e.g. ros-melodic-xacro) but it did not. Had to add apt-get installs later again. (https://github.com/Field-Robotics-Lab/glider_hybrid_whoi/blob/c1b3aff1b45be5935b931e32933af7dec21f5b8d/docker/Dockerfile#L160-L182)
3. Can you provide more details about sourcing environment scripts other than writing it at `.bashrc`

Could you give it a look? It's at a new branch `Kinematics_only_newdockerfile`
https://github.com/Field-Robotics-Lab/glider_hybrid_whoi/blob/c1b3aff1b45be5935b931e32933af7dec21f5b8d/docker/Dockerfile#L1-L201

daewok committed:
Change a dependency that looked to be named incorrectly

daewok commented:
I just pushed some commits to the new branch. With them, I'm able to do `./build.bash` and `./run.bash -w glider_kinematics` and get the simulator to run. (note that I'm not using my GPU and I haven't tried with a GPU yet).

The reason so many deps were missed is that the `rosdep ... install` command was actually failing, but it was masked since it was in a pipe and `-o pipefail` was not set. I changed that so it should now complain if rosdep fails for some reason.

woensug-choi commented:
> I just pushed some commits to the new branch. With them, I'm able to do `./build.bash` and `./run.bash -w glider_kinematics` and get the simulator to run. (note that I'm not using my GPU and I haven't tried with a GPU yet).
> 
> The reason so many deps were missed is that the `rosdep ... install` command was actually failing, but it was masked since it was in a pipe and `-o pipefail` was not set. I changed that so it should now complain if rosdep fails for some reason.

Can't launch anything from my side. I am getting X error for `gazebo` at `./run.bash -w glider_kinematics`. Any suggestions on how to solve this issue?

woensug-choi commented:
@mabelzhang Isn't this a magic line to set model paths? I am trying to reference the models newly included in this PR.
```
  <export>
    <gazebo_ros gazebo_model_path="${prefix}/models"/>
    <gazebo_ros gazebo_media_path="${prefix}"/>
  </export>
```
I am trying to upload a Buzzard Bay heightmap. But I can't locate it using `model://`. It shouldn't be difficult. What else do I have to do to add a path for models in this PR?

mabelzhang commented:
That looks correct. Adding the path usually takes me a few rounds of back and forth testing...
Are you installing the packages when you compile? Try `rospack find` to see where ROS thinks the package is, whether in `src` or in `devel` or `install`. It could be that the models are only in `src` but `rospack find` is locating a compiled directory. In that case, you would want to install the models directory in CMakeLists.txt.
You could run Gazebo with --verbose to see if it prints anything about the path.

woensug-choi committed:
Introducing Buzzard bay partially

woensug-choi commented:
### Added buzzards bay partially (Quicks Hole area) Near with automatic bathymetry respawning plugin.
-  **Launch command**
   ```bash
   roslaunch glider_hybrid_whoi_gazebo BuzzBay.launch
   ```
- Click the `play` button on the Gazebo window and wait for the first bathymetry to be spawned.
- Next bathymetry tile will be spawned and the previous tile will be removed automatically.

***

### Problmes to be solved
- [x] : **Spherical coordinate system in the simulator**
    - Not 100% confident that the bathymetry tiles are correctly oriented.
        - Intended glider initial location
          (Coordinate translated from EPSG:4326 to EPSG:26987)
          ```
          Lat / Lon : -70.853° / 41.45°
          X / Y : 470503.30 / 50036.53
          ```
        - Tiles are correctly spawned according to the intended initial location
          ```bash
          [Dbg] [bathy_plugin.cc:396] Inserted model: buzzbay.R_-70.857_-70.852_41.447_41.452
          ```
    - GPS sensor data is generating unexpected Lat/Lon values.
        - GPS sensor definition doesn't take any much arguments for the coordinate system. I was guessing it's taking parent links (base_link) frame which has ned transform. (On rqt window, base_link has transformation OK)
        - Other coordinate system settings is at `Buzzbay.world` inherited from Dave project's plugin which also inherited from Mike Jakuba's plugin from Wind_energy. Regretfully, not sure what it does exactly.
           ```xml
           <!-- Setting this to origin of EPSG:26987 -->
           <spherical_coordinates>
              <longitude_deg>-70.5</longitude_deg>
              <latitude_deg>41.0</latitude_deg>
           </spherical_coordinates>
           ```
        - Current result of the GPS sensor data is as follows (Looks like ENU/NED coordinate frame error)
           ```
           longitude: 76.0371042964
           latitude: -40.4165916257
           ```
    - Possible solutions
      - [x] Fix launch/xacro files defining coordinate frames.
      - [ ] Insert a code to calculate correct Lat/Lon for final output of GliderStatus.msg
    - Related
      - https://github.com/Field-Robotics-Lab/dave/issues/77
      - https://github.com/uuvsimulator/uuv_simulator/issues/165

woensug-choi committed:
fixed lat/lon value (accuracy issue remains), full Buzzbay included

woensug-choi commented:
### Added FULL Buzzards bay bathymetry
- Roughly 1500x1500 m tiles with 50 m overlap regions are included (almost 780 MB)
-  **Launch command**
   ```bash
   roslaunch glider_hybrid_whoi_gazebo BuzzBay.launch
   ```
- Click the `play` button on the Gazebo window and wait for the first bathymetry to be spawned.
- **Testing script for auto-spawning and auto-removing of tiles**
   ```bash
   rosrun direct_kinematics_ros_plugins test_directKinematics.py
   ```
   - The script contains moving the glider toward east (`rudder_control_mode=RUDDER_CONTROL_HEADING=1` with `target_heading = pi/2`)
   - Next bathymetry tile will be spawned and the previous tile will be removed automatically.

### Issue on GPS accuracy
The vehicle's position and GPS value do not match exactly.
- Vehicle's initial position
  - `x: 480381.43`, `y: 56666.85` which is `Lon: -70.735`, `Lat: 41.51` calculated using [epsg.io](https://epsg.io/transform#s_srs=4326&t_srs=26987&x=-70.7350000&y=41.5100000)
- GPS value obtained from UwGliderStatus.msg
  - `Lon: -70.712083444`, `Lat: 41.2136645438`

This is also the case of Dave repo's rexrov tutorial case https://github.com/Field-Robotics-Lab/dave/issues/77#issuecomment-748878031. It seems it's a problem with the GPS sensor since the bathymetry is spawning correctly.

woensug-choi commented:
I think the PR is ready to review for merge. @daewok could you review this from the kinematics functionality point of view?

Future Todo
- Include stratified ocean current (with Kolumbo data)
- Separate GPS sensor location and the actual location
  - the actual location (x,y (EPSG:26987 NAD83) from converted into lat/lon (EPSG:4326 WGS 84))

Future issues
- Docker deployments: Dockerfile satisfying both Eric's usage and X window for Gazebo
- GPS sensor accuracy

daewok commented:
I'm just getting back online after moving and xmas. I also have my
desktop with GPU finally set up. So I think I'm going to tackle the
Docker stuff first, since having it Dockerized will make my workflow
easier.

woensug-choi commented:
Happy new year! Were you able to get the dockerfile working? Did you review this PR? I only see the Merge button.

daewok committed:
gazebo: install worlds directory

daewok committed:
Massively rework the Docker images

They should be much more cache efficient now, perform correctly (always build
the code by COPYing it in instead of using git clone), and are based on
upstream ROS images.

There are additionally two images: a dev one that requires code to be compiled
in the image before use and a prod one that has everything compiled already.

See the README for more details.

daewok committed:
gazebo: install models folder

daewok committed:
Split out "Demo-y" components of BuzzBay launch into BuzzBayDemo

Want to have some way to easily launch the bare necessities (which does not
include rviz or joystick control).

daewok committed:
Update Dockerfiles

Now that bathymetry is included, we need the latest version of
Dave. Additionally, it exposed various issues with libraries not declaring
dependencies correctly or not installing all their files.

See:

https://bitbucket.org/whoidsl/ds_sim/pull-requests/5/install-all-libraries

https://github.com/Field-Robotics-Lab/dave/pull/88

https://github.com/Field-Robotics-Lab/nps_uw_sensors_gazebo/pull/42

daewok committed:
Add first version of docker-compose file

Launches both the simulator (in buzzards bay) and the simulated glider control
system.

daewok committed:
Buffer output in compose, print a bit less in run bash scripts

daewok committed:
Improve Dockerfiles for use without buildkit

There might be some weirdness with args. There's definitely a difference with
COPY --chown=... Buildkit will chown all directories it creates, legacy builder
will only chown the files and directories it actually transfers.

daewok committed:
Fix some bugs with directories in Docker build scripts

daewok commented:
Just pushed my work. I tried to update the docker/README.MD file well enough to
document how it works now.

I also added a docker-compose.yml file that launches both the Gazebo simulator
and the simulated glider control system. You should just need to run
`docker-compose build` and `docker-compose up` to have the simulator start
up. In the absence of any other input, the glider control system just tries to
make the glider safe by surfacing.

The docker-compose file currently assumes that if you want to use a GPU
the nvidia runtime is set as your default Docker runtime. If that's not the case,
you'll need to create a local docker-compose.override.yml file setting the runtime
to use.

To get a glider console, you can run:

    docker-compose exec glider_control_sim /ros_entrypoint.sh rosrun slocum_glider_sim_driver slocum_glider_sim_console

You can then run any of the subset of Glider commands that the simulated
control system supports. You can run the simple one dive mission with `run
ini0.mi`. You can run a simple box mission with `run inibox01.mi`. After the
missions complete, you'll need to hit Ctrl-C to tell it to finish. If you want
to quit the console hit Ctrl-D.

I did notice two oddities:

1. Sometimes Gazebo just hangs on me. Time stops progressing in the GUI and no
   more messages are sent from the plugins. I noticed this occasionally outside
   of Docker as well. Have you seen this?
2. The messages sent by the plugins have a time stamp of over 1600000000
   seconds. That makes me wonder if the /use_sim_time parameter is being set
   early enough. I don't remember if this happened to me earlier or not.

Take a look at the commit message for 92fac83. It'd be nice if we could get
those PRs mentioned in it merged before this is merged.

woensug-choi committed:
Fixed build.bash and use_sim_time

woensug-choi commented:
Lots of commits! I am starting to understand why we need this. I've dealt with all PRs on this and dave repository.
- Dev/Prod image: works as expected!
- Docker compose: I get X Error.
```bash
gazebo_1              | libGL error: No matching fbConfigs or visuals found
gazebo_1              | libGL error: failed to load driver: swrast
gazebo_1              | X Error of failed request:  GLXBadContext
gazebo_1              |   Major opcode of failed request:  152 (GLX)
gazebo_1              |   Minor opcode of failed request:  6 (X_GLXIsDirect)
gazebo_1              |   Serial number of failed request:  44
gazebo_1              |   Current serial number in output stream:  43
```

The last commit includes,
- Exclude pandas python library for test_directKinematics.py script
- Fix target dockerfile of `build.bash`
- Fix to set `/use_sim_time` parameter correctly at BuzzBay.launch file
- Set writeLog to false

For two oddities,
- No I have not experienced any hang-ups on Gazebo. Sometimes the gazebo complains about missing material. But I've never experienced freezing. Maybe not enough space at /tmp? (Changed writeLog to false)
- I've fixed the sim time problem! They were supposed to be in `paused=false`, `use_sim_time=true`, `gui=true` order. I've found an article that some codes should be added to make ROS node to wait for time synchronization [article](https://answers.gazebosim.org//question/3172/timing-synchronisation-between-gazebo-and-ros/). I did not. But it works fine.

woensug-choi committed:
Fixed build.bash and use_sim_time

daewok committed:
Use the merged versions of dependencies

daewok commented:
Weird. It worked on my desktop with an nvidia GPU and I just tested it on the computer in Caryn House mostly successfully (`docker-compose build` doesn't work on that computer... I'm guesing it's because the version of docker-compose on there is ancient and has a bug. I had to build with `build.bash` and then use `docker-compose up`).

Is the nvidia runtime set as the default on the computer you were using?

woensug-choi commented:
@daewok How do I check if the Nvidia runtime is set as the default?

Brian has suggested that we merge this PR if the kinematics functionality is working before we diverge too much in one PR. I think it is very reasonable to do so. This PR already includes Kinematics, Bathymetry, and Docker. What do you think?

daewok commented:
Run `docker info` and it should be in the output. If it's not the default, you can typically add `"default-runtime": "nvidia"` to `/etc/docker/daemon.json` (or however your OS configures Docker). You should also be able to use a docker-compose override file to tell docker-compose to use nvidia instead of the default, but I'm not sure what the contents of that should be. (And I don't think there's any issue with making nvidia the default over runc).

I wholeheartedly agree about merging this soon. It's already a beast.

woensug-choi commented:
I messed up my system. Now back online. The workstation at WHOI has nvidia runtime as default. I'll double check on my local machine.

woensug-choi committed:
Another fix to use_sim_time rosparam (moved declaration to the very top in the launch file)

woensug-choi commented:
### Docker
- Runs ok with the following commands 👍 
```bash
# Terminal Window 1 at docker folder
build.bash
run.bash
roslaunch glider_hybrid_whoi_gazebo BuzzBay.launch
```
- Rosparam `use_sim_time` is working correctly and could be checked by following commands 👍 
```bash
# Terminal Window 2 at docker folder
join.bash
source /opt/ros/melodic/setup.bash
source ~/glider_hybrid_whoi/install/setup.bash
rostopic echo /glider_hybrid_whoi/direct_kinematics/UwGliderStatus
```
The log at the first window still shows time stamp of over 1600000000, however, the UwGliderStatus time stamp read on the rostopic correctly shows sim_time value (which is calculated using ros::time::now())

### Docker-compose
- With Nvidia runtime, `build.bash` and then use `docker-compose up` launched ok 👍 
```bash
# Terminal Window 1 at docker folder
build.bash
docker-compose up
```
- Mission terminal console works as it's supposed to be 👍 
```bash
# Terminal Window 2 at docker folder
docker-compose exec glider_control_sim /ros_entrypoint.sh rosrun slocum_glider_sim_driver slocum_glider_sim_console
# Ctrl+C to see the typable terminal cursor
run ini0.mi
run inibox01.mi
```
Is there `join.bash` functionality for docker-compose? I wanted to check whether `use_sim_time` works correctly. 

During the tests, I have not seen any Gazebo hang-ups. Currently, building on another machine without Nvidia runtime. Will see what happens. (Update: with runc docker runtime, it works perfectly 👍 ).

**I think this PR is ready to merge now.**

daewok commented:
> Is there join.bash functionality for docker-compose? I wanted to check whether use_sim_time works correctly.

Something like this is probably what you want:

    docker-compose exec gazebo bash

woensug-choi commented:
I was able to call up a new window using `docker-compose exec glider_control_sim /ros_entrypoint.sh bash`
The use_sim_time was correctly working in the docker-compose environment too.
